Arrowgene.Ddon.GameServer/Handler/DispelExchangeDispelItemHandler.cs

```diff
@@ -0,0 +1,154 @@
+using System.Collections.Generic;
+
+namespace Arrowgene.Ddon.GameServer.Handler
+{
+    public class DispelExchangeDispelItemHandler : GameRequestPacketHandler<C2SDispelExchangeDispelItemReq, S2CDispelExchangeDispelItemRes>
+    {
+        private static readonly ServerLogger Logger = LogProvider.Logger<ServerLogger>(typeof(DispelExchangeDispelItemHandler));
+
+        public DispelExchangeDispelItemHandler(DdonGameServer server) : base(server)
+        {
+        }
+
+        public override S2CDispelExchangeDispelItemRes Handle(GameClient client, C2SDispelExchangeDispelItemReq request)
+        {
+            var res = new S2CDispelExchangeDispelItemRes();
+
+            var appraisialItems = Server.AssetRepository.SpecialShopAsset.AppraisalItems;
+            foreach (var item in request.GetDispelItemList)
+            {
+                if (!appraisialItems.ContainsKey(item.Id))
+                {
+                    // TODO: Throw an exception?
+                    continue;
+                }
+
+                bool toBag = false;
+                switch (item.StorageType)
+                {
+                    case 19:
+                        toBag = true;
+                        break;
+                    case 20:
+                        toBag = false;
+                        break;
+                    default:
+                        throw new Exception($"Unexpected destination when exchanging items {item.StorageType}");
+                }
+
+                // Check for cost
+                bool hasFullPayment = true;
+                foreach (var payment in item.UIDList)
+                {
+                    hasFullPayment &= Server.ItemManager.HasItem(Server, client.Character, ItemManager.BothStorageTypes, payment.UId, payment.Num);
+                }
+
+                if (!hasFullPayment)
+                {
+                    // Probably a hacker or cheater
+                    continue;
+                }
+
+                var updateCharacterItemNtc = new S2CItemUpdateCharacterItemNtc()
+                {
+                    UpdateType = ItemNoticeType.GetDispelItem
+                };
+
+                // Consume payment
+                foreach (var payment in item.UIDList)
+                {
+                    List<CDataItemUpdateResult> updateResults = Server.ItemManager.ConsumeItemByUIdFromMultipleStorages(Server, client.Character, ItemManager.BothStorageTypes, payment.UId, payment.Num);
+                    updateCharacterItemNtc.UpdateItemList.AddRange(updateResults);
+                }
+
+                var purchase = AppraiseItem(client.Character, appraisialItems[item.Id]);
+
+                List<CDataItemUpdateResult> itemUpdateResults = Server.ItemManager.AddItem(Server, client.Character, toBag, purchase.ItemId, purchase.ItemNum);
+
+                var newItem = client.Character.Storage.FindItemByUIdInStorage(ItemManager.BothStorageTypes, itemUpdateResults[0].ItemList.ItemUId).Item2.Item2;
```

Comment:
If the purchase.ItemNum is over the stack limit, the item will be split across multiple stacks, meaning it'd be two instances, each with different UIDs. If we expect this to ever happen we should do this in a loop over itemUpdateResults, if not, we should assert that the itemUpdateResults list count is of one item